### PR TITLE
New article about applying live effects to images and videos

### PR DESCRIPTION
**Pull request by wibblymat**

### CODE REVIEW

File: src/content/en/fundamentals/media/manipulating/live-effects.md [hunk 1]
```diff
@@ -0,0 +1,376 @@
...
+Anyone who has worked with WebGL before will probably realise that I can't go into the full detail
+of how this works in a single article. If you'd like to learn more about WebGL, check out the
+recommended reading at the end of this article.
+
+However, you can take a general overview of what needs to be done in the case of manipulating a
+single image.
+
+One of the most important things to remember about WebGL is that it is **not** a 3D graphics API. In
+fact, WebGL (and OpenGL) is good at exactly one thing - drawing triangles. In your application you
+must work out describe what you *actually* want to draw in terms of triangles. In the case of a 2D
```
Comment (jpmedley) [inline src/content/en/fundamentals/media/manipulating/live-effects.md]: Is it 'work out' or 'describe'. 'Describe' is a stronger verb, but it may not be an appropriate description.

File: src/content/en/fundamentals/media/manipulating/live-effects.md [hunk 2]
```diff
@@ -0,0 +1,376 @@
...
+
+<img src="images/quad-from-triangles.png" class="attempt-right" />
+
+The basic process is:
+- Send data to the GPU that describes the vertices (points) of the triangles
+- Send your source image to the GPU as a texture (image)
+- Create a 'vertex shader'
+- Create a 'fragment shader'
+- Set some shader variables, called 'uniforms'
+- Run the shaders
```
Comment (jpmedley) [inline src/content/en/fundamentals/media/manipulating/live-effects.md]: If they must happen in this order, then use numbers instead of bullets.

While you're there, add periods to the ends. These are complete sentences.

File: src/content/en/fundamentals/media/manipulating/live-effects.md [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+
```
Comment (jpmedley) [inline src/content/en/fundamentals/media/manipulating/live-effects.md]: Add something like, "Let's go into detail."

File: src/content/en/fundamentals/media/manipulating/live-effects.md [hunk 4]
```diff
@@ -0,0 +1,376 @@
...
+You allocate some memory on the graphics card called a vertex buffer, and you store data in it that
```
Comment (jpmedley) [inline src/content/en/fundamentals/media/manipulating/live-effects.md]: Use 'Start by allocating'.

This and the previous comment strengthen the forward flow of the article.

File: src/content/en/fundamentals/media/manipulating/live-effects.md [hunk 5]
```diff
@@ -0,0 +1,376 @@
...
+Pretty much every kind of 2D image manipulation that you might want to do can be done in the
+fragment shader, and all of the other WebGL parts can be abstracted away. You can see
+[the abstraction layer](https://github.com/GoogleChrome/snapshot/blob/master/src/image-shader.ts)
+(in TypeScript) that is being in used in one of our sample applications if you'd like to see the
+details.
+
+### Which should I use?
+For pretty much any professional quality image manipulation, you should use WebGL. There is no
+getting away from the fact that this kind of work is the whole reason GPUs were invented. You can
+process images much, much faster on the GPU, which is essential for any real-time effects.
```
Comment (jpmedley) [inline src/content/en/fundamentals/media/manipulating/live-effects.md]: 'much, much' could be replaced with something stronger or more descriptive. I don't actually know how much faster a GPU is, but let's say for the sake of argument that it is orders of magnitude faster. I would actually say 'orders of magnitude' instead of 'much, much'.

File: src/content/en/fundamentals/media/manipulating/live-effects.md [hunk 6]
```diff
@@ -0,0 +1,376 @@
...
+However, there are plenty of abstractions around for WebGL that mean you can get the performance
+boost without needing to learn the details.
+
+If even if you parallelize your code with `Worker`s, your GPU may have 100s of times as many
+specialized cores as your CPU has general cores.
+
+Examples in this article are mostly for 2D canvas, but the principles should translate pretty easily
+to fragment shader code.
+
+## Effect Types
```
Comment (jpmedley) [inline src/content/en/fundamentals/media/manipulating/live-effects.md]: Sentence case for all but the article title. Please fix throughout.

File: src/content/en/fundamentals/media/manipulating/live-effects.md [hunk 7]
```diff
@@ -0,0 +1,376 @@
...
+
+### Pixel Effects
+This is the simplest category to both understand and implement. All of these transformations take
+the color value of a single pixel and passes it into a function that will give you back another
+color value to use as the output.
```
Comment (jpmedley) [inline src/content/en/fundamentals/media/manipulating/live-effects.md]: Change 'that will give you back another color value to use as the output' to 'that returns another color value'.

File: src/content/en/fundamentals/media/manipulating/live-effects.md [hunk 8]
```diff
@@ -0,0 +1,376 @@
...
+
+There are many variations on these operations that are more or less complicated. Some will take into
+account how the human brain processes visual information based on decades of research, and some will
+be dead simple ideas that give an effect that's mostly reasonable. Either way, the idea is the same.
```
Comment (jpmedley) [inline src/content/en/fundamentals/media/manipulating/live-effects.md]: Which is what, exactly?

File: src/content/en/fundamentals/media/manipulating/live-effects.md [hunk 9]
```diff
@@ -0,0 +1,376 @@
...
+    const brightness = 1.1; // Make the image 10% brighter
+    for (let i = 0; i < imageData.data.length; i += 4) {
+      imageData.data[i] = imageData.data[i] * brightness;
+      imageData.data[i + 1] = imageData.data[i + 1] * brightness;
+      imageData.data[i + 2] = imageData.data[i + 2] * brightness;
+    }
+
+Note that the loop moves 4 bytes along at a time, but only changes three values - this is because it
+skips changing the alpha value. Also remember that a Uint8ClampedArray will round all values to
+integers, and clamp any value greater than 255.
```
Comment (jpmedley) [inline src/content/en/fundamentals/media/manipulating/live-effects.md]: Then how does multiplying by 1.1 make anything brigher? I'm missing something.

File: src/content/en/fundamentals/media/manipulating/live-effects.md [hunk 10]
```diff
@@ -0,0 +1,376 @@
...
+
+This is easy enough, though. When you initially create your imageData object you can make a copy of
+the data.
+
+    const originalPixels = new Uint8Array(imageData.data);
+
+Note that in the WebGL case it isn't actually possible to query the **new** color of another pixel
+from inside the shader, so you always get the original.
+
+The most common category of multi-pixel effect is called a convolution filter. You take each nearby
```
Comment (jpmedley) [inline src/content/en/fundamentals/media/manipulating/live-effects.md]: "You take" make it sound as though the web page developer is doing this manually. I think you mean "The convolution filter takes."

If you make that change be sure to fix subject-verb agreement throughout the sentence.

File: src/content/en/fundamentals/media/manipulating/live-effects.md [hunk 11]
```diff
@@ -0,0 +1,376 @@
...
+the current pixel. For example, this is the kernel for a 3x3 Gaussian blur.
+
+    | 0  1  0 |
+    | 1  4  1 |
+    | 0  1  0 |
+
+So to calculate the output pixel you take the colors of the pixels above, below, to the left and to
+the right and sum them all together with four times the value of the current pixel. Then you divide
+the result by 8. You can see how the result will be a pixel that is mostly the original, but with
+the nearby pixels bleeding in.
```
Comment (jpmedley) [inline src/content/en/fundamentals/media/manipulating/live-effects.md]: So the only pixel you're changing is the center? Do I have that right? You should show the result somehow.

**Comment by jpmedley** [comment]
@wibblymat My typo fixes are usually nothing serious. I changed one paragraph I felt needed work. You should take a look at that.

File: src/content/en/fundamentals/media/manipulating/live-effects.md [hunk 12]
```diff
@@ -0,0 +1,374 @@
+project_path: /web/_project.yaml
+book_path: /web/fundamentals/_book.yaml
+description: Many of today's most popular apps let you apply filters and effects to images or video. This article talks about how to implement these features on the open web.
+
+{# wf_updated_on: 2017-08-08 #}
+{# wf_published_on: 2017-08-08 #}
+{# wf_blink_components: Blink>WebGL, Blink>Canvas #}
+
+# Live effects for images and video {: .page-title }
```
Comment (jpmedley) [inline src/content/en/fundamentals/media/manipulating/live-effects.md]: This should be title case.

Comment (jpmedley) [inline src/content/en/fundamentals/media/manipulating/live-effects.md]: It also needs to match the toc.

File: src/content/en/fundamentals/media/manipulating/live-effects.md [hunk 13]
```diff
@@ -0,0 +1,374 @@
...
+talks about how to implement these features on the open web.
+
+The process is basically the same for videos and images, but I'll cover some important
+video considerations at the end. Throughout the article you can assume that 'image' means 'image a or
+single frame of a video'
+
+## How to get at the pixel data for an image
+
+There are 3 basic categories of image manipulation that are common:
+- Pixel effects like contrast, brightness, warmth, sepia tone, saturation.
```
Comment (surma) [inline src/content/en/fundamentals/media/manipulating/live-effects.md]: Needs a new line before this to show as a proper list.

File: src/content/en/fundamentals/media/manipulating/live-effects.md (hunk 12 above)
Comment (surma) [inline src/content/en/fundamentals/media/manipulating/live-effects.md]: I don’t like the title that much, honestly. I think it’s the work “live”. Maybe rather “real-time”?

File: src/content/en/fundamentals/media/manipulating/live-effects.md [hunk 14]
```diff
@@ -0,0 +1,374 @@
...
+    const imageData = context.getImageData(0, 0, canvas.width, canvas.height);
+    const pixels = imageData.data;
+
+At this point the `pixels` variable is a `Uint8ClampedArray` with a length of `width * height * 4`.
+Every array element is one byte and every four elements in the array represents the color of one
+pixel. Each of the four elements represents amount of red, green, blue, and alpha (transparency)
+in that order. The pixels are ordered starting from the top-left corner and working left to right
+and top to bottom.
+
+    pixels[0] = red value for pixel 0
```
Comment (surma) [inline src/content/en/fundamentals/media/manipulating/live-effects.md]: This is being syntax highlighted weirdly. @petele Can we prevent syntax highlighting on blocks somehow?

File: src/content/en/fundamentals/media/manipulating/live-effects.md [hunk 15]
```diff
@@ -0,0 +1,374 @@
...
+think of. However, this array is a *copy* of the actual pixel data for the canvas. To write the
+edited version back you need to use the `putImageData` method to write this back to the top-left
+corner of the canvas
+
+    context.putImageData(imageData, 0, 0);
+
+### WebGL
+Anyone who has worked with WebGL before will probably realise that I can't go into the full detail
+of how this works in a single article. If you'd like to learn more about WebGL, check out the
+recommended reading at the end of this article.
```
Comment (surma) [inline src/content/en/fundamentals/media/manipulating/live-effects.md]: Anchor link?

File: src/content/en/fundamentals/media/manipulating/live-effects.md [hunk 16]
```diff
@@ -0,0 +1,374 @@
...
+
+One of the most important things to remember about WebGL is that it is **not** a 3D graphics API. In
+fact, WebGL (and OpenGL) is good at exactly one thing - drawing triangles. In your application you
+must work out describe what you *actually* want to draw in terms of triangles. In the case of a 2D
+image, that's very simple, because a rectangle is two similar right-angled triangles, arranged so
+that their hypotenuses are in the same place.
+
+<img src="images/quad-from-triangles.png" class="attempt-right" />
+
+The basic process is:
```
Comment (surma) [inline src/content/en/fundamentals/media/manipulating/live-effects.md]: Same. Needs newline

File: src/content/en/fundamentals/media/manipulating/live-effects.md [hunk 17]
```diff
@@ -0,0 +1,374 @@
...
+- Send data to the GPU that describes the vertices (points) of the triangles
+- Send your source image to the GPU as a texture (image)
+- Create a 'vertex shader'
+- Create a 'fragment shader'
+- Set some shader variables, called 'uniforms'
+- Run the shaders
+
+You allocate some memory on the graphics card called a vertex buffer, and you store data in it that
+describes each point of each triangle. You can also set some variables, called uniforms, that are
+global values through both shaders.
```
Comment (surma) [inline src/content/en/fundamentals/media/manipulating/live-effects.md]: I’m a bit torn about this paragraph. On the one hand you say “you can’t go into details because OpenGL lol”, but then you throw technical terms in here. If you know OpenGL, you know how to render a quad. If you don’t, these words mean nothing to you and don’t add any value. 

It’s a thin line and I don’t know where the right balance is, but I’d lean towards removing the technical terms, explaining the process more in layman’s terms and linking to a more in-depth explanation (do a follow up article maybe?)

File: src/content/en/fundamentals/media/manipulating/live-effects.md [hunk 18]
```diff
@@ -0,0 +1,374 @@
...
+
+A vertex shader uses data from the vertex buffer to calculate where on the screen to draw the three
+points of a triangle.
+
+Now the GPU knows which pixels within the canvas need to be drawn. The fragment shader is called
+once per pixel, and needs to return the color that should be drawn to the screen. The fragment
+shader can read information from one or more textures to determine the color.
+
+When reading a texture in a fragment shader you specify which part of the image you want to read
+using two floating-point coordinates between 0 (left or bottom) and 1 (right or top).
```
Comment (surma) [inline src/content/en/fundamentals/media/manipulating/live-effects.md]: Maybe add an image?

File: src/content/en/fundamentals/media/manipulating/live-effects.md [hunk 19]
```diff
@@ -0,0 +1,374 @@
...
+    uniform vec2 textureSize;
+    uniform sampler2D textureSampler;
+
+    main() {
+      vec2 textureCoords = pixelCoords / textureSize;
+
+      gl_FragColor = texture2D(textureSampler, textureCoords);
+
+      // ...
+    }
```
Comment (surma) [inline src/content/en/fundamentals/media/manipulating/live-effects.md]: It might not be obvious to some reader that this is fragment shader code.

File: src/content/en/fundamentals/media/manipulating/live-effects.md [hunk 20]
```diff
@@ -0,0 +1,374 @@
...
+Pretty much every kind of 2D image manipulation that you might want to do can be done in the
+fragment shader, and all of the other WebGL parts can be abstracted away. You can see
+[the abstraction layer](https://github.com/GoogleChrome/snapshot/blob/master/src/image-shader.ts)
+(in TypeScript) that is being in used in one of our sample applications if you'd like to see the
+details.
+
+### Which should I use?
+For pretty much any professional quality image manipulation, you should use WebGL. There is no
+getting away from the fact that this kind of work is the whole reason GPUs were invented. You can
+process images much, much faster on the GPU, which is essential for any real-time effects.
```
Comment (surma) [inline src/content/en/fundamentals/media/manipulating/live-effects.md]: But _why_ is it faster? “Because one thread per pixel blabla”.

File: src/content/en/fundamentals/media/manipulating/live-effects.md [hunk 21]
```diff
@@ -0,0 +1,374 @@
...
+There are many variations on these operations that are more or less complicated. Some will take into
+account how the human brain processes visual information based on decades of research, and some will
+be dead simple ideas that give an effect that's mostly reasonable. Either way, the idea is the same.
+
+For example, a brightness control can be implemented by simply taking the red, green and blue values
+of the pixel and multiplying them by a brightness value. A brightness of 0 will make the image
+entirely black. A value of 1 will leave the image unchanged. A value greater than 1 will make it
+brighter.
+
+For 2D canvas:
```
Comment (surma) [inline src/content/en/fundamentals/media/manipulating/live-effects.md]: Could it be interesting to showcase code for _both_ WebGL and 2D for comparison

File: src/content/en/fundamentals/media/manipulating/live-effects.md [hunk 22]
```diff
@@ -0,0 +1,374 @@
...
+This slightly changes how you do things in the 2D canvas case because you want to be able to read
+the *original* colors of the image, and the previous example was updating the pixels in place.
+
+This is easy enough, though. When you initially create your image data object you can make a copy
+of the data.
+
+    const originalPixels = new Uint8Array(imageData.data);
+
+Note that in the WebGL case it isn't actually possible to query the **new** color of another pixel
+from inside the shader, so you always get the original.
```
Comment (surma) [inline src/content/en/fundamentals/media/manipulating/live-effects.md]: I stumbled over this. Maybe phrase it from another POV: “Note that you can only access the input image in WebGL, as the output value of other pixels might not have been calculated yet.” or something

File: src/content/en/fundamentals/media/manipulating/live-effects.md [hunk 23]
```diff
@@ -0,0 +1,374 @@
...
+
+The most common category of multi-pixel effect is called a convolution filter. You take each nearby
+pixel, including the current one, and add some multiple of the color, called the weight, to a total.
```
Comment (surma) [inline src/content/en/fundamentals/media/manipulating/live-effects.md]: Maybe simplify a bit? “A convolution filter is a weighted sum of a pixel and its neighbors” and link out to convolution filter on Wikipedia.

File: src/content/en/fundamentals/media/manipulating/live-effects.md [hunk 24]
```diff
@@ -0,0 +1,374 @@
...
+
+        const outputIndex = (x + y * imageWidth) * 4;
+        imageData.data[outputIndex] = redTotal / weightTotal;
+        imageData.data[outputIndex + 1] = greenTotal / weightTotal;
+        imageData.data[outputIndex + 2] = blueTotal / weightTotal;
+      }
+    }
+
+This gives the basic idea, but there are many guides out there that will go into much more
+detail, and list many other useful kernels.
```
Comment (surma) [inline src/content/en/fundamentals/media/manipulating/live-effects.md]: Any specific article that you recommend/would link to?

File: src/content/en/fundamentals/media/manipulating/live-effects.md [hunk 25]
```diff
@@ -0,0 +1,374 @@
...
+        sx, sy, sw, sh,
+        dx, dy, dw, dh);
+
+Rotation and reflection are directly available in the 2D context. Before you draw the image into the
+canvas, change the various transforms.
+
+    // Move the canvas so that the center of the canvas is on the Y-axis
+    context.translate(-canvas.width / 2, 0);
+
+    // An X scale of -1 reflects the canvas in the Y-axis
```
Comment (surma) [inline src/content/en/fundamentals/media/manipulating/live-effects.md]: “inverts”?

File: src/content/en/fundamentals/media/manipulating/live-effects.md (hunk 2 above)
Comment (wibblymat) [inline src/content/en/fundamentals/media/manipulating/live-effects.md]: Not strictly ordered. The last item must be last, but the first two can happen anywhere before that. Setting uniforms must be after creating the shaders, but the shaders can be done either way around.

File: src/content/en/fundamentals/media/manipulating/live-effects.md (hunk 25 above)
Comment (wibblymat) [inline src/content/en/fundamentals/media/manipulating/live-effects.md]: Perhaps either word works from a technical perspective, but the effect is a reflection.

File: src/content/en/fundamentals/media/manipulating/live-effects.md (hunk 17 above)
Comment (wibblymat) [inline src/content/en/fundamentals/media/manipulating/live-effects.md]: I've changed the tone slightly. Instead of saying 'this is too difficult', I say 'here is an introduction'. So for people who don't know OpenGL the idea is to teach them roughly what these terms mean.

**Comment by wibblymat** [comment]
@jpmedley @surma Thanks for the reviews! I *think* I resolved everything here.

File: src/content/en/fundamentals/media/manipulating/live-effects.md (hunk 25 above)
Comment (surma) [inline src/content/en/fundamentals/media/manipulating/live-effects.md]: Your call, really. I just found it ambiguous: The word “reflects” could be seen to mean “convey”.
